### ISSUE TITLE
[Win10] LWSWS doesn't seems to find plugins in given plugin path

### ISSUE THREAD

**Issue by alexgille** [warmcat/libwebsockets#540]
Hi @lws-team! (Now my turn to bother you haha)

I'm trying to implement an LWSWS on **Windows 10** with **Visual Studio 2015**. Compilation is fine and executable runs correctly, but it doesn't seems to load my protocol plugin and use it while trying to connect to it.

* What's working
Running compiled server, connect and get web content (html, js, img)

* What's not working
Placing the example **protocol_dumb_increment.dll** plugin in "./plugins" directory
Connecting to "ws://localhost/" with "dumb-increment-protocol" as subprotocol
Make libwebsockets HTML testing example work

**main.c - Copied and adapted from 'lwsws/main.c'** :
```
#include "stdafx.h"
#include <signal.h>

#define LWSWS_CONFIG_STRING_SIZE (32 * 1024)

struct lws_context * context;

void signal_cb(uv_signal_t *watcher, int signum)
{
	lwsl_err("Signal %d caught, exiting...\n", watcher->signum);

	switch (watcher->signum) {
		case SIGTERM:
		case SIGINT:
			break;
		default:
			signal(SIGABRT, SIG_DFL);
			abort();
			break;
	}

	lws_libuv_stop(context);
}

static const char * const plugin_dirs[] = {
	"./plugins",
	NULL
};

int _tmain(int argc, _TCHAR* argv[])
{
	struct lws_context_creation_info info;
	int opts = 0, cs_len = LWSWS_CONFIG_STRING_SIZE - 1;
	char config_str[LWSWS_CONFIG_STRING_SIZE];
	char * cs = config_str;

	lws_set_log_level(0xFFFF, NULL);

	lwsl_notice("libwebsockets test server - license LGPL2.1+SLE\n");
	lwsl_notice("(C) Copyright 2010-2016 Andy Green <andy@warmcat.com>\n");
	lwsl_notice(" Using resource path \"%s\"\n", plugin_dirs[0]);

	memset(&info, 0, sizeof info);
	info.max_http_header_pool = 32;
	info.options = opts | LWS_SERVER_OPTION_EXPLICIT_VHOSTS | LWS_SERVER_OPTION_VALIDATE_UTF8 | LWS_SERVER_OPTION_LIBUV;
	info.plugin_dirs = plugin_dirs;
	info.timeout_secs = 15;

	if (lwsws_get_config_globals(&info, "./", &cs, &cs_len)) {
		lwsl_err("unable to get config globals\n");
		return false;
	}

	context = lws_create_context(&info);
	if (context == NULL) {
		lwsl_err("libwebsocket init failed\n");
		return false;
	}

	if ( ! lwsws_get_config_vhosts(context, &info, "./", &cs, &cs_len)) {

		lws_uv_sigint_cfg(context, 1, signal_cb);
		lws_uv_initloop(context, NULL, 0);

		lws_libuv_run(context, 0);
	}

	lws_context_destroy(context);

	return TRUE;
}
```

**vhost - localhost**
```
{
    "vhosts": [{
        "name": "localhost",
        "port": "80",
        "mounts": [{
            "mountpoint": "/",
            "origin": "file://../WebSources/",
            "default": "index.html"
        }],
		"ws-protocols": [{
			"dumb-increment-protocol": {
				"default": "1"
			}
		}]
    }]
}
```

**console output**

> C:\dev\WebDisplayCtrl\Release>WebDisplayConsole.exe
> [2016/04/25 10:00:04:6909] NOTICE: libwebsockets test server - license LGPL2.1+SLE
> [2016/04/25 10:00:04:6909] NOTICE: (C) Copyright 2010-2016 Andy Green <andy@warmcat.com>
> [2016/04/25 10:00:04:6909] NOTICE:  Using resource path "./plugins"
> [2016/04/25 10:00:04:7066] NOTICE: Initial logging level 65535
> [2016/04/25 10:00:04:7066] NOTICE: Libwebsockets version: 2.0.0 desktop-kevdfjd\cm3@DESKTOP-KEVDFJD-
> [2016/04/25 10:00:04:7066] NOTICE: IPV6 not compiled in
> [2016/04/25 10:00:04:7066] NOTICE: libev support not compiled in
> [2016/04/25 10:00:04:7066] NOTICE: libuv support compiled in and enabled
> [2016/04/25 10:00:04:7066] NOTICE:  Threads: 1 each 30000 fds
> [2016/04/25 10:00:04:7066] NOTICE:  Compiled with OpenSSL support
> [2016/04/25 10:00:04:7222] NOTICE:  mem: per-conn:          648 bytes + protocol rx buf
> [2016/04/25 10:00:04:7222] NOTICE:  canonical_hostname = DESKTOP-KEVDFJD
> [2016/04/25 10:00:04:7222] NOTICE:   adding protocol dumb-increment-protocol
> [2016/04/25 10:00:04:7222] NOTICE: Creating Vhost 'localhost' port 80, 1 protocols
> [2016/04/25 10:00:04:7222] NOTICE:    mounting file://../WebSources/ to /
> [2016/04/25 10:00:04:7222] NOTICE:  Using non-SSL mode
> [2016/04/25 10:00:04:8941] NOTICE: lws_protocol_init
> [2016/04/25 10:00:11:1834] ERR: 127.0.0.1 - - [25/May/2016:10:00:11 +0200] "GET /index.html http/1.1" 200 18698 Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.102 Safari/537.36
> [2016/04/25 10:00:11:3794] ERR: 127.0.0.1 - - [25/May/2016:10:00:11 +0200] "GET /libwebsockets.org-logo.png http/1.1" 200 7196 Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.102 Safari/537.36
> [2016/04/25 10:00:27:0215] NOTICE: wsi 0230D5B8: TIMEDOUT WAITING on 3 (did hdr 0, ah 00CEBA28, wl 0, pfd events 0)
> [2016/04/25 10:00:27:0215] NOTICE: lws_header_table_detach: wsi 0230D5B8: ah held 16s, ah.rxpos 0, ah.rxlen 0, mode/state 0 4,wsi->more_rx_waiting 0
> [2016/04/25 10:01:12:2482] NOTICE: lws_header_table_detach: wsi 0230BF80: ah held 61s, ah.rxpos 0, ah.rxlen 0, mode/state 0 4,wsi->more_rx_waiting 0
> [2016/04/25 10:01:48:1308] ERR: 192.168.63.1 - - [25/May/2016:10:01:48 +0200] "GET / http/1.1" 200 18698
> [2016/04/25 10:01:48:1308] NOTICE: error on reading from skt : 10053
> [2016/04/25 10:08:10:5509] ERR: Signal 2 caught, exiting...
> [2016/04/25 10:08:10:5509] NOTICE: lws_context_destroy

Thanks !

**Comment by lws-team** [comment]
Sure, thanks for the help earlier.

Just a general comment, I don't think the ./ and ../ type filepaths are a good idea.  These should be elaborated into absolute paths I think, at least the ones in config (html can play that game better).  Otherwise it'll break / act different depending on the working directory where you started it.

When it works right, during early boot you should see something like this

```
lwsws[19005]:  mem: platform fd map: 800000 bytes
lwsws[19005]:   Plugins:
lwsws[19005]:   Scanning /usr/share/libwebsockets-test-server/plugins/
lwsws[19005]:    libprotocol_client_loopback_test.so
lwsws[19005]:    libprotocol_dumb_increment.so
lwsws[19005]:    libprotocol_generic_sessions.so
lwsws[19005]:    libprotocol_lws_mirror.so
lwsws[19005]:    libprotocol_lws_server_status.so
lwsws[19005]:    libprotocol_lws_status.so
lwsws[19005]:    libprotocol_lwsct.so
lwsws[19005]:    libprotocol_post_demo.so
```

so comparing with your log, the basic problem is it's not even looking for them.

It looks like only the unix plat stuff initializes it, even though the actual business end of it is in libuv.c.  So I pushed a patch

https://github.com/warmcat/libwebsockets/commit/5c27334b39f1f5f280db007de451483bbb71677a

 to also have the windows plat stuff init it, see what happens with that.

**Comment by alexgille** [comment]
Hi,

Is SqLite3 required for this version? I have the following CMake issue :

**Command Line**
```
cmake -G "Visual Studio 14"
	-DOPENSSL_ROOT_DIR=C:\tools\build_openssl-1.0.2h
	-DOPENSSL_INCLUDE_DIR=C:\tools\build_openssl-1.0.2h\include
	-DOPENSSL_LIBRARIES=C:\tools\build_openssl-1.0.2h\lib\libeay32.lib;C:\tools\build_openssl-1.0.2h\lib\ssleay32.lib
	-DLWS_WITH_LWSWS=ON
	-DLIBUV_INCLUDE_DIRS=C:\tools\build_libuv-v1.8.0\include
	-DLIBUV_LIBRARIES=C:\tools\build_libuv-v1.8.0\libuv.lib
	-DCMAKE_INSTALL_PREFIX=C:\tools\build_libwebsockets-2.0.0
	..
```

I checked with `cmake .. -LH` and found a **LWS_WITH_SQLITE3** flag so I tried generating with explicit `-DLWS_WITH_SQLITE3=OFF` but cmake is not taking it into account.

**CMake console output**
```
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
SQLITE3_INCLUDE_DIRS
   used as include directory in directory C:/tools/libwebsockets-2.0.0
   [...]
   used as include directory in directory C:/tools/libwebsockets-2.0.0
SQLITE3_LIBRARIES
    linked by target "websockets" in directory C:/tools/libwebsockets-2.0.0
    linked by target "websockets_shared" in directory C:/tools/libwebsockets-2.0.0
```

**Comment by lws-team** [comment]
Yes atm lwsws temporarily implies sqlite3, due to the new generic-sessions plugin I debugging with someone else atm.

You can get rid of the dependency by editing this bit near the top of CMakeLists.txt

if (LWS_WITH_LWSWS)
 message(STATUS "LWS_WITH_LWSWS --> Enabling LWS_WITH_PLUGINS and LWS_WITH_LIBUV")
 set(LWS_WITH_PLUGINS 1)
 set(LWS_WITH_LIBUV 1)
 set(LWS_WITH_ACCESS_LOG 1)
 set(LWS_WITH_SERVER_STATUS 1)
 set(LWS_WITH_LEJP 1)
 set(LWS_WITH_GENERIC_SESSIONS 1)   <<<<==== kill this
endif()
```

**Comment by alexgille** [comment]
Compilation is fine, lwsws runs et scans for plugins. I put the dumb increment protocol plugin only, but this happends at runtime :

```
[2016/04/25 15:33:08:0921] ERR: Failed to get init on protocol_dumb_increment.dll: The specified procedure could not be found.
[2016/04/25 15:33:08:1011] ERR: Failed to get init on protocol_dumb_increment.dll: error: 127
```

**Comment by lws-team** [comment]
It males some assumptions about the plugin filenames vs function names.

I pushed a patch that adapts the assumptions for win32 case.

9de9d0dacd4a53fa639af0b4ab1eb30c3e76c6e5

**Comment by alexgille** [comment]
I still have the same error messages.
I manually added a **printf** to print the computed function name, and it was fine.
Is it normal that **LWS_VISIBLE** has no value on Windows? Maybe symbols are not public or exported...

**Comment by alexgille** [comment]
Another bad news: HTTP protocol doesn't work neither with this version...

**Comment by lws-team** [comment]
Well the last patch fixes the function name generation, which was wrong, so we're headed in the right direction.

Can you try adding

`#define LWS_VISIBLE __declspec(dllexport)`

at the top of the protocol .c file?

**Comment by lws-team** [comment]
Looking at it closer, there's already a way in libwebsockets.h to get that defined.for win32.

In current master

 - I removed the association of LWS_WITH_GENERIC_SESSIONS from LWS_WITH_LWSWS, so you don't need to worry about sqlite3

 - I added the couple of lines of magic to each plugin so LWS_VISIBLE will be correct now for windows, you don't need to add it by hand

So just updating to master is worth testing.

**Comment by alexgille** [comment]
What's on master now still needs to remove LWS_WITH_GENERIC_SESSIONS automatically set to 1 from CMakeList, and unfortunately, I still have the same plugin loading issue...

**Comment by alexgille** [comment]
The `#define LWS_VISIBLE __declspec(dllexport)` trick didn't work neither

**Comment by alexgille** [comment]
Manually replacing `LWS_VISIBLE` by `__declspec(dllexport)` ends up with **protocol_dumb_increment.lib** and **protocol_dumb_increment.exp** being generated in **build/lib/Release**

**Comment by lws-team** [comment]
You're right, I must have dreamed I pushed it.

Try 4d501bb6994959d6d3ba4ace0e2857be0e7895c5.

About your last post, the #define and "manually replacing it" must be the same if the #define came after the #include of libwebsockets.h.

What does it mean we got a ..lib, it's a static library?   Previously it was generating a .dll?

**Comment by alexgille** [comment]
Thanks.

About the **.lib**, it's an extra **.lib** that's generated in addition with the **.dll**. I'm quiet sure that it's the dll-loading static library.

**Comment by lws-team** [comment]
OK then if it still fails to bind to the init function, please have a look with dumpbin on the dll to see what the exports are actually called (maybe decorated symbols?)

**Comment by alexgille** [comment]
I don't really know how to use dumpbin but I tried :

**dumpbin /EXPORTS protocol_dumb_increment.dll**
```
Microsoft (R) COFF/PE Dumper Version 14.00.23918.0
Copyright (C) Microsoft Corporation.  All rights reserved.


Dump of file protocol_dumb_increment.dll

File Type: DLL

  Summary

        1000 .data
        1000 .gfids
        1000 .rdata
        1000 .reloc
        1000 .rsrc
        2000 .text
```

Same output with **/SYMBOLS**

**Comment by alexgille** [comment]
Good news !

I managed to get symbols exported by using
`extern __declspec(dllexport)`
instead of
`__declspec(dllexport)`

It still generates an extra useless `protocol_dumb_increment.lib` but now it works.

There's still a more important problem, now HTTP doesn't work at all with the given vhost configuration.

**dumpbin output**
```
$> dumpbin /EXPORTS protocol_dumb_increment.dll
Microsoft (R) COFF/PE Dumper Version 14.00.23918.0
Copyright (C) Microsoft Corporation.  All rights reserved.


Dump of file protocol_dumb_increment.dll

File Type: DLL

  Section contains the following exports for protocol_dumb_increment.dll

    00000000 characteristics
    5746C54B time date stamp Thu May 26 11:43:39 2016
        0.00 version
           1 ordinal base
           2 number of functions
           2 number of names

    ordinal hint RVA      name

          1    0 00001240 destroy_protocol_dumb_increment
          2    1 00001250 init_protocol_dumb_increment

  Summary

        1000 .data
        1000 .gfids
        1000 .rdata
        1000 .reloc
        1000 .rsrc
        2000 .text
```

**Console output for my server** 
```
[...]
[2016/04/26 11:45:19:7657] NOTICE:   Plugins:
[2016/04/26 11:45:19:7667] NOTICE:   Scanning ./plugins
[2016/04/26 11:45:19:7687] NOTICE:    protocol_dumb_increment.dll
[2016/04/26 11:45:19:7819] NOTICE:   Scanning C:/dev/WebDisplayCtrl/Release/plugins
[2016/04/26 11:45:19:7836] NOTICE:    protocol_dumb_increment.dll
[2016/04/26 11:45:19:7851] NOTICE:  Compiled with OpenSSL support
[...]
```

**Comment by lws-team** [comment]
Well... okay... microsoft....  I'll try implement this in the sources.

From your logs, the plugin scan and import is working now, so that is good.

So what does "not work at all" mean, and what exactly is the vhost configuration?  Can you paste the lwsws config?

**Comment by JoakimSoderberg** [comment]
> It still generates an extra useless protocol_dumb_increment.lib but now it works.

Please note that `.lib` on windows is not the same as `.a` or `.so` on Unix.

> Dynamic libraries on Windows have the .dll extension; in order to link against a DLL, an import library is required. The import library is a static library. It contains the code required to load the DLL.

taken from this stackoverflow question: http://stackoverflow.com/questions/6422478/linking-a-lib-and-def-files

.... But maybe you already knew that, I did not read through this entire thread :)

**Comment by alexgille** [comment]
* **What's working**
Loading plugin dll
Connecting from Javascript to websocket server with dumb-increment-protocol

* **What's not working** (sorry for confusion, it's not HTTP that's not working, but non-WS HTTP)
Getting web files (HTML, CSS, JS) (404 Error). The same vhost configuration was working on previous commits of libwebsocket

**conf**
```
{
	"global": {
		"count-threads": "1",
		"server-string": "lwsws",
		"init-ssl": "yes",
		"plugin-dir": "C:/dev/WebDisplayCtrl/Release/plugins"
	}
}
```

**conf.d/localhost**
```
{
    "vhosts": [{
        "name": "localhost",
        "port": "80",
        "mounts": [{
            "mountpoint": "/",
            "origin": "file://C:/dev/WebDisplayCtrl/WebSources/",
            "default": "index.html"
        }],
		"ws-protocols": [{
			"dumb-increment-protocol": {
				"default": "1"
			}
		}]
    }]
}
```

**Comment by alexgille** [comment]
@JoakimSoderberg thanks for your comment, but I already knew the difference between static and dynamic libraries for Unix and Windows. By "useless .lib" I meant "a static library that I didn't need and that was created anyway". It looks like Visual Studio automatically generates the static dll-loading library when you export symbols. I think it's a default Visual Studio / Win32 behaviour.

**Comment by lws-team** [comment]
I think I captured the necessary change in the plugins to make the export behaviour and updated master.

There's only two ways to get a 404

One is it tried and failed to open the file, he should show the log

	if (wsi->u.http.fd == LWS_INVALID_FILE) {
		lwsl_err("Unable to open '%s'\n", file);
		lws_return_http_status(wsi, HTTP_STATUS_NOT_FOUND, NULL);

The other is because we ended up at the default protocol handler, he does nothing just returns a 404 because no mount served that URL.

Are you seeing the log about "Unable to open"?  Is the path sane if so?

**Comment by lws-team** [comment]
Hmmm aren't you going to get caught out by this kind of thing?

```
	snprintf(path, sizeof(path) - 1, "%s/%s", origin, uri);
```

Is windows smart enough to translate / to \ nowadays or we need to deal with it?

**Comment by alexgille** [comment]
**[*]** Here is my complete console output after running the server and refreshing the page couple times :

```
[2016/04/26 11:53:46:4338] NOTICE: Initial logging level 65535
[2016/04/26 11:53:46:4348] NOTICE: Libwebsockets version: 2.0.0 desktop-kevdfjd\cm3@DESKTOP-KEVDFJD-
[2016/04/26 11:53:46:4378] NOTICE: IPV6 not compiled in
[2016/04/26 11:53:46:4388] NOTICE: libev support not compiled in
[2016/04/26 11:53:46:4438] NOTICE: libuv support compiled in and enabled
[2016/04/26 11:53:46:4458] NOTICE:  Threads: 1 each 30000 fds
[2016/04/26 11:53:46:4508] NOTICE:   Plugins:
[2016/04/26 11:53:46:4518] NOTICE:   Scanning ./plugins
[2016/04/26 11:53:46:4568] NOTICE:    protocol_dumb_increment.dll
[2016/04/26 11:53:46:4588] NOTICE:   Scanning C:/dev/WebDisplayCtrl/Release/plugins
[2016/04/26 11:53:46:4628] NOTICE:    protocol_dumb_increment.dll
[2016/04/26 11:53:46:4648] NOTICE:  Compiled with OpenSSL support
[2016/04/26 11:53:46:4702] NOTICE:  mem: per-conn:          656 bytes + protocol rx buf
[2016/04/26 11:53:46:4732] NOTICE:  canonical_hostname = DESKTOP-KEVDFJD
[2016/04/26 11:53:46:4752] NOTICE: libwebsockets test server - license LGPL2.1+SLE
[2016/04/26 11:53:46:4772] NOTICE: (C) Copyright 2010-2016 Andy Green <andy@warmcat.com>
[2016/04/26 11:53:46:4802] NOTICE:  Using resource path "./plugins"
[2016/04/26 11:53:46:4822] NOTICE:   adding protocol dumb-increment-protocol
[2016/04/26 11:53:46:4872] NOTICE: Creating Vhost 'localhost' port 80, 3 protocols
[2016/04/26 11:53:46:4882] NOTICE:    mounting http://C:/dev/WebDisplayCtrl/WebSources/ to /
[2016/04/26 11:53:46:4936] NOTICE:  Using non-SSL mode
[2016/04/26 11:53:46:6207] NOTICE: lws_protocol_init
[2016/04/26 11:53:46:6214] NOTICE:     vh localhost prot dumb-increment-protocol opt default
[2016/04/26 11:53:46:6224] NOTICE: Setting default protocol for vh localhost to dumb-increment-protocol
[2016/04/26 11:53:46:6244] NOTICE:     vh localhost prot dumb-increment-protocol opt default
[2016/04/26 11:53:46:6310] NOTICE: Setting default protocol for vh localhost to dumb-increment-protocol
// ===========================
[2016/04/26 11:53:52:7094] NOTICE: no hit
[2016/04/26 11:53:52:7134] ERR: 127.0.0.1 - - [26/May/2016:11:53:52 +0200] "GET /index.html http/1.1" 404 161 Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.102 Safari/537.36
[2016/04/26 11:53:53:5244] NOTICE: no hit
[2016/04/26 11:53:53:5254] ERR: 127.0.0.1 - - [26/May/2016:11:53:53 +0200] "GET /index.html http/1.1" 404 161 Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.102 Safari/537.36
[2016/04/26 11:53:54:0620] NOTICE: no hit
[2016/04/26 11:53:54:0630] ERR: 127.0.0.1 - - [26/May/2016:11:53:54 +0200] "GET /index.html http/1.1" 404 161 Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.102 Safari/537.36
[2016/04/26 11:53:54:6525] NOTICE: no hit
[2016/04/26 11:53:54:6535] ERR: 127.0.0.1 - - [26/May/2016:11:53:54 +0200] "GET /index.html http/1.1" 404 161 Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.102 Safari/537.36
[2016/04/26 11:53:54:8952] NOTICE: no hit
[2016/04/26 11:53:54:8972] ERR: 127.0.0.1 - - [26/May/2016:11:53:54 +0200] "GET /index.html http/1.1" 404 161 Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.102 Safari/537.36
[2016/04/26 11:53:55:0690] NOTICE: no hit
[2016/04/26 11:53:55:0701] ERR: 127.0.0.1 - - [26/May/2016:11:53:55 +0200] "GET /index.html http/1.1" 404 161 Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.102 Safari/537.36
[2016/04/26 11:54:11:6338] NOTICE: wsi 00BDF6F0: TIMEDOUT WAITING on 3 (did hdr 0, ah 005C0C00, wl 0, pfd events 0)
[2016/04/26 11:54:11:6367] NOTICE: lws_header_table_detach: wsi 00BDF6F0: ah held 16s, ah.rxpos 0, ah.rxlen 0, mode/state 0 4,wsi->more_rx_waiting 0
```

Outputs after the '===' line were printed after refreshing the page. Each refresh will print these lines :

```
[2016/04/26 11:53:55:0690] NOTICE: no hit
[2016/04/26 11:53:54:8972] ERR: 127.0.0.1 - - [26/May/2016:11:53:54 +0200] "GET /index.html http/1.1" 404 161 Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.102 Safari/537.36
```

I've never seen the 'no hit' message before, I think it means that the server was not able to "hit" the page.

**[*]** The `/` path seperator is handled on Windows for POSIX-compliancy reasons. Here is an explanation from [MSDN](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247(v=vs.85).aspx#paths).

> Note:  File I/O functions in the Windows API convert "/" to "\" as part of converting the name to an NT-style name, except when using the "\\?\" prefix as detailed in the following sections.

**Comment by lws-team** [comment]
"No hit" means it thinks there was no mount with a mountpoint that matched.

This is bad in the log

[2016/04/26 11:53:46:4882] NOTICE:    mounting http://C:/dev/WebDisplayCtrl/WebSources/ to /

The protocol type is converted to an ordinal from a list, it's got it wrong somewhere...file:// has become http://

Good news about the / ....

**Comment by lws-team** [comment]
I don't see how that can process the mount without printing something in the log

./lwsws/conf.c

```
		static const char * const mount_protocols[] = {
			"http://",
			"https://",
			"file://",
			"cgi://",
			">http://",
			">https://",
			"callback://"
		};

...

		for (n = 0; n < ARRAY_SIZE(mount_protocols); n++)
			if (!strncmp(a->m.origin, mount_protocols[n],
			     strlen(mount_protocols[n]))) {
				lwsl_err("----%s\n", a->m.origin);
				m->origin_protocol = n;
				m->origin = a->m.origin +
					    strlen(mount_protocols[n]);
				break;
			}

		if (n == ARRAY_SIZE(mount_protocols)) {
			lwsl_err("unsupported protocol:// %s\n", a->m.origin);
			return 1;
		}
```

It should either blow the "unsupported protocol" or print the "---- file://" in the log, but it doesn't show either in your log.

Is this tree latest master?

**Comment by alexgille** [comment]
Now it works, I have web content and web socket connection. The example is working fine too.
My project was using a previous version of **conf.c** and **c headers**, my bad...

You were right, it doesn't really like the "./plugins" path notation.

I wonder why there is this message at each HTTP GET :
`ERR: 127.0.0.1 - -`

**output**
```
[2016/04/26 14:38:59:3606] NOTICE:  Using resource path "./plugins"
[2016/04/26 14:38:59:3656] ERR: ----file://C:/dev/WebDisplayCtrl/WebSources/
[2016/04/26 14:38:59:3676] NOTICE:   adding protocol dumb-increment-protocol
[2016/04/26 14:38:59:3716] NOTICE: Creating Vhost 'localhost' port 80, 3 protocols
[2016/04/26 14:38:59:3736] NOTICE:    mounting file://C:/dev/WebDisplayCtrl/WebSources/ to /
[2016/04/26 14:38:59:3766] NOTICE:  Using non-SSL mode
[2016/04/26 14:38:59:4910] NOTICE: lws_protocol_init
[2016/04/26 14:38:59:4910] NOTICE:     vh localhost prot dumb-increment-protocol opt default
[2016/04/26 14:38:59:4939] NOTICE: Setting default protocol for vh localhost to dumb-increment-protocol
[2016/04/26 14:38:59:4969] NOTICE:     vh localhost prot dumb-increment-protocol opt default
[2016/04/26 14:38:59:4989] NOTICE: Setting default protocol for vh localhost to dumb-increment-protocol
[2016/04/26 14:39:26:9847] ERR: 127.0.0.1 - - [26/May/2016:14:39:26 +0200] "GET /index.html http/1.1" 200 18698 Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.102 Safari/537.36
```

**Comment by lws-team** [comment]
Great.

1) Maybe.... I was imagining people would just use lwsws directly and put everything else in plugins.  But I guess it might not fit every case... if so that might make sense.

2) It's not that it "doesn't like" ./ it's that it makes operation depend on your cwd.  If you run this as, eg, a service, or give it to some other unsuspecting guy to run, the cwd used when it is spawned may be uncontrolled.  So it's making it fragile.

3) Those messages are the Apache-compatible logs.  If you define a filepath in the vhost part of the conf like this

"access-log": "/var/log/lwsws/test-access-log",

they will go in there without the lwsl logging preamble instead of the main log, and you can use them with Apache type log analyzers.

**Comment by alexgille** [comment]
Thanks for the help and hints !

I'm actually trying to make a software embedding a WebSocket and HTTP server with a GUI, working as a middleware like a "binary-to-JSON message translator".

At the end my program will run the server based on lwsws example on one side, connect to another software that communicates with a custom binary protocol on the other side, in order to give a JSON translation of those custom messages to a web front-end for presentation.

Today, the **lwsws/main.c** code is adapted and wrapped in C++ class in my project and works fine. Except for the blocking call to `lws_libuv_run`, I might run my server in a separate thread; I don't know yet.

**Comment by alexgille** [comment]
I forgot to tell you, `LWS_WITH_GENERIC_SESSIONS` keep requiring SQLITE3, even when disabled in CMake call

**Comment by lws-team** [comment]
OK... I'll look at moving lwsws/conf.c

Cmake caches the options, you'll need to do `cmake .. -DLWS_WITH_GENERIC_SESSION=0` or nuke and regenerate your build dir to convince it otherwise... now it simply doesn't set it, it doesn't force it off so the old forced-on setting remains cached by cmake.  It's not really an lws thing.

**Comment by alexgille** [comment]
Alright. Thanks for your help and I hope all of this helped improve your lib !

**Comment by alexgille** [comment]
**CMakeLists.txt** from master keeps acting like this, even when generating on a freshly-unzipped project or when using `-DLWS_WITH_GENERIC_SESSIONS=0` option on CMake call.

The **CMakeLists.txt** I have here contains :
```
if (LWS_WITH_LWSWS)
 message(STATUS "LWS_WITH_LWSWS --> Enabling LWS_WITH_PLUGINS and LWS_WITH_LIBUV")
 set(LWS_WITH_PLUGINS 1)
 set(LWS_WITH_LIBUV 1)
 set(LWS_WITH_ACCESS_LOG 1)
 set(LWS_WITH_SERVER_STATUS 1)
 set(LWS_WITH_LEJP 1)
 set(LWS_WITH_GENERIC_SESSIONS 1)   <--
endif()
```

**Comment by lws-team** [comment]
Yeah... what I said is true but it doesn't help if the line forcing it set gets left in :-P

I just pushed an update to master where 

 - GENERIC_SESSIONS is really no longer forced on by LWSWS
 - ./lwsws/conf.c -> ./lib/lejp-conf.c
 - ./lwsws/lwsws.h is no longer needed and is deleted
 - CMake has a new LWS_WITH_LEJP_CONF, which is set by LWS_WITH_LWSWS
 - the 2 apis from ./lwsws/conf.c are exposed in libwebsockets.h

```
LWS_VISIBLE LWS_EXTERN int
lwsws_get_config_globals(struct lws_context_creation_info *info, const char *d,
			 char **config_strings, int *len);

LWS_VISIBLE LWS_EXTERN int
lwsws_get_config_vhosts(struct lws_context *context,
			struct lws_context_creation_info *info, const char *d,
			char **config_strings, int *len);
```

So you should be able to remove your copy of conf.c and get the apis from lws.  lwsws is happy with it (and is only 5.3KB now on x86_64...)

**Comment by alexgille** [comment]
Awesome! Many thanks for your reactivity and your work :smile:
